smart-contracts/test/Unlock/migrateLockData.js

```diff
@@ -0,0 +1,368 @@
+      assert.equal(await unlock.publicLockLatestVersion(), pastVersion + 1)
+      assert.equal(await lock.publicLockVersion(), pastVersion + 1)
+    })
+
+    it('schemaVersion still undefined', async () => {
+      assert.equal((await lock.schemaVersion()).toNumber(), 0)
+    })
+
+    it('relevant records have been updated', async () => {
+      for (let i = 0; i < 100; i++) {
+        const tokenId = i + 1
+        assert.equal(await lock.isValidKey(tokenId), true)
+        assert.equal(await lock.ownerOf(tokenId), keyOwners[i].address)
+        assert.equal(await lock.balanceOf(keyOwners[i].address), 1)
+        assert.equal(await lock.getHasValidKey(keyOwners[i].address), true)
+        assert.equal(
+          (await lock.keyExpirationTimestampFor(tokenId)).toNumber(),
+          expirationTimestamps[i].toNumber()
+        )
+      }
+    })
+
+    it('rest of the records have NOT been updated', async () => {
+      const totalSupply = (await lock.totalSupply()).toNumber()
+      for (let i = 100; i < totalSupply; i++) {
+        const tokenId = i + 1
+        assert.equal(await lock.isValidKey(tokenId), false)
+        assert.equal(await lock.ownerOf(tokenId), keyOwners[i].address) // ownerOf should work regardless
+        assert.equal(await lock.balanceOf(keyOwners[i].address), 0)
+        assert.equal(await lock.getHasValidKey(keyOwners[i].address), false)
+        assert.equal(
+          (await lock.keyExpirationTimestampFor(tokenId)).toNumber(),
+          0
+        )
+      }
+    })
+
+    describe('features for key are deactivated', () => {
+      let someBuyers
+      beforeEach(async () => {
+        const signers = await ethers.getSigners()
+        someBuyers = signers.slice(1, 5)
+      })
+
+      it('purchase should fail ', async () => {
+        await reverts(
+          lock.connect(someBuyers[0]).purchase(
+            [],
+            someBuyers.map((k) => k.address),
+            someBuyers.map(() => web3.utils.padLeft(0, 40)),
+            someBuyers.map(() => web3.utils.padLeft(0, 40)),
+            [],
+            {
+              value: (keyPrice * keyOwners.length).toFixed(),
+            }
+          ),
+          'MIGRATION_REQUIRED'
+        )
+      })
+
+      it('grantKeys should fail ', async () => {
+        await reverts(
+          lock.connect(someBuyers[0]).grantKeys(
+            someBuyers.map((k) => k.address),
+            someBuyers.map(() => Date.now()),
+            someBuyers.map(() => web3.utils.padLeft(0, 40))
+          ),
+          'MIGRATION_REQUIRED'
+        )
+      })
+      it('extend should fail ', async () => {
+        const [, generousBuyer] = await ethers.getSigners()
+        await reverts(
+          lock
+            .connect(generousBuyer)
+            .extend(0, 1, web3.utils.padLeft(0, 40), [], {
+              value: keyPrice,
+            }),
+          'MIGRATION_REQUIRED'
+        )
+      })
+    })
+
+    describe('relaunch remaining data migration', () => {
```

Comment:
I want to test when we have to call this function twice... (or 3 times if we count the migration)

Reply:
Ok I added the tests for this